azkaban-common/src/main/java/azkaban/executor/ExecutionControllerUtils.java

```diff
@@ -367,6 +418,39 @@ public static void failEverything(final ExecutableFlow exFlow, final Status fina
       }
     }
 
+    // Fire correct JOB_STARTED and JOB_FINISHED events.
+    if (eventHandler != null && projectManager != null) {
+      Project project = projectManager.getProject(exFlow.getProjectId());
+      Flow flow = project.getFlow(exFlow.getFlowId());
+      for (Entry<ExecutableNode, Status> entry: node2origStatus.entrySet()) {
+        ExecutableNode node = entry.getKey();
+        Status origStatus = entry.getValue();
+
+        // Fill in job props by following the logic in ProjectManagerServlet.
+        if (node.getInputProps() == null) {
+          node.setInputProps(projectManager.getJobOverrideProperty(project, flow, node.getId(),
```

Comment:
Why reading from file here? As discussed earlier, this is already in ImmutableFlowProps

Reply:
ImmutableFlowProps is not persisted into memory but gets GC after uploading a project zip (in transient DirectoryYamlFlowLoader and DirectoryFlowLoader objects). And it cannot cover “job override properties” overridden via api call.

I think my current implementation is fine. It is referring to the design as in this page: https://<clusterName>/manager?project=<projectName>&flow=<flowName>&job=<jobName>. (specifically ProjectManagerServlet.java)